File: study1/week1/merge-nodes-in-between-zeros.java
// https://leetcode.com/problems/merge-nodes-in-between-zeros/
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    // Time Complexity: O(n)
    // Space Compexity: O(n)
    public ListNode mergeNodes(ListNode head) {
        ListNode result = new ListNode();
        ListNode iterator = result;
        int temp = 0;

        while (head != null) {
            if (head.val == 0 && temp != 0) {
                iterator.val = temp;
                temp = 0;
                if (head.next != null) {    // last ListNode
                    iterator.next = new ListNode();
                    iterator = iterator.next;
                }
            } else if (head.val != 0) {
                temp += head.val;
            }
            head = head.next;
        }

        return result;
    }
}

File: study1/solutions/week1/merge_nodes_in_between_zeros.py
# https://leetcode.com/problems/merge-nodes-in-between-zeros/

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
       self.val = val
       self.next = next

class Solution:
    def mergeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        current_head = head.next
        merged_head = ListNode()
        merged = merged_head
        prev_merged = None

        while current_head != None:
            if current_head.val == 0:
                merged.next = ListNode()
                prev_merged = merged
                merged = merged.next
            else:
                merged.val += current_head.val
            current_head = current_head.next
        prev_merged.next = None
        return merged_head
            

